## Eine Einführung in Python und Jupyter Notebooks

Dieses Jupyter Notebook ist das erste in einer Reihe von Notebooks die das Ziel verfolgen mathematische Inhalte aus der Vorlesung mit Programmieraufgaben zu festigen und gegebenenfalls zu erweitern. Gleichzeitig wird die Programmiersprache Python den Studierenden von Null auf über die Länge des Kurses beigebracht. Konkret wird in diesem Notebook die Notwendigkeit von Programmierkenntnissen motiviert, sowie eine Einführung in die Grundlagen von Python und Jupyter Notebooks geboten.

Inhalt:
<ul>
 <li><a href="#C1">1. Grundlegende Rechenoperationen und Variablen</a></li>
 <li><a href="#C2">2. Listen</a></li>
 <li><a href="#C3">3. Schleifen</a></li>
 <li><a href="#C4">4. Funktionen</a></li>
 <li><a href="#C5">5. Texte in Python: strings</a></li>
 <li><a href="#C6">6. if-Abfragen</a></li>
 <li><a href="#C7">7. Lösungen</a></li>
</ul>


**Python Grundlagen:** Keine <br>
**Math. Grundlagen:** Abiturniveau

<div style = "padding: 15px; border-radius: 8px; border: 2px solid; border-color: Green; background-color: #E0FFD1;">
<p> Das Dateiformat, das Sie hier sehen nennt sich <b>Jupyter Notebook</b> (.ipynb). Solche Notebooks werden dazu verwendet Computercodes mithilfe von Fließtexten, mathematischen Formeln, Bildern und interaktiven Widgets anschaulicher zu machen. Wir werden Jupyter in diesem Kurs dazu verwenden Ihnen die Grundlagen der Programmiersprache Python beizubringen und Ihnen parallel anwendungsbezogene Programmieraufgaben zu stellen. </p>

Hier einige Hinweise zur Verwendung der Jupyter Notebooks:
 - Ein Jupyter Notebook setzt sich aus Zellen bzw. Blöcken zusammen, von denen es zwei Typen gibt: **Textblöcke** und **Codeblöcke**. Textblöcke enthalten Aufgabenstellungen und Erklärtexte und müssen somit nicht von Ihnen verändert werden. Codeblöcke enthalten Pythoncode, der entweder vorgegeben ist oder von Ihnen selbst geschrieben werden muss.
 - Zum Ausführen eines Blocks klicken Sie auf *Start* (▶) in der oberen Leiste.
 - Möchten Sie ein laufendes Programm in einem Codeblock abbrechen klicken Sie auf *Stopp*. (⏹)
 - In der oberen Zeile befinden sich noch weitere Schaltflächen mit denen das Notebook editiert werden kann. Fährt man mit dem Mauszeiger über ein Symbol, wird eine Erklärung zu diesem angezeigt. Für die Verwendung unserer Notebooks werden außer *Start* und *Stopp* allerdings keine weiteren Schaltflächen benötigt.
 - Alle Codeblöcke sind miteinander verbunden. Sie müssen diese also der Reihe nach (von oben nach unten) ausführen. Die Zahl die links neben einem ausgeführten Codeblock steht gibt die Reihenfolge an in der der Codeblock ausgeführt wurde.
 - Haben Sie vorgefertigte Text- oder Codeblöcke verändert und möchten diese Änderungen rückgängig machen, verwenden Sie die Tastenkombination *Strg + Z*. Diese müssen Sie so häufig anwenden bis alle Ihre Änderungen einzeln rückgängig gemacht wurden.

### I. Motivation

Viele Probleme in Naturwissenschaft und Technik lassen sich nicht von Hand lösen. Hierzu zählen nicht nur komplexe Gleichungssysteme und Integrale, sondern auch die Auswertung von Experimenten, deren Datensätze häufig mehrere zehntausend (oder mehr!) Datenpunkte beinhalten. Ein Werkzeug, welches bei diesen und auch vielen weiteren Problemen Abhilfe schafft, ist die Programmiersprache **Python**. Sie ist eine der beliebtesten und am weitest verbreiteten Programmiersprachen der Welt. Dies liegt nicht nur daran, dass sie vergleichsweise einfach zu erlernen ist, sondern auch daran, dass sie trotz ihrer Einfachheit sehr vielseitig einsetzbar ist. Von der Darstellung einfacher Graphen bishin zu Simulationen und neuronalen Netzwerken, Python ist häufig die erste Wahl. Grundlagenkenntnisse in Python sind damit in vielen Bereichen unverzichtbar und diese bereits im Studium zu erlernen ist wichtig. 

Hier einige konkrete Beispiele für Probleme bzw. Aufgabenstellungen, welche ohne Computerhilfe kaum zu bewältigen sind:

#### I.1. Beispiel:  Numerische Integration

Möchte man den Umfang einer Ellipse berechnen so muss zwangsweise das sogenannte **elliptische Integral** $E(k, \phi)$ gelöst werden. Es gilt

$$U = 4a \cdot E(\varepsilon, \pi /2) = 4a\int_0^{\pi / 2} \mathrm{d}\theta \sqrt{1 - \varepsilon^2 \sin^2 \theta} \ , \quad \varepsilon = \sqrt{1 - (b/a)^2} \ , \quad a \geq b \ .$$

> **Am Rande:** Ähnliche Integrale treten unter anderem auch bei der Berechnung von Kepplerschen Planetenbahnen auf. 

Es ist nicht möglich das oben auftretende Integral analytisch (= von Hand) zu lösen, was zur Folge hat, dass es keine simple Formel für den Ellipsenumfang geben kann. Stattdessen muss man auf computergestützte Verfahren zurückgreifen welche sich mit hoher Präzision an das exakte numerisches Ergebnis annähern können.

#### I.2. Beispiel: Auswertung von Experimenten

Betrachten wir ein Experiment in dem der Ton einer schwingenden Gitarrensaite aufgenommen wird. Das Mikrofon ist in der Lage den  durch die schwingende Saite erzeugten Schalldruck mit einer Zeitauflösung von $\sim 20 \mathrm{\mu s}$ aufzunehmen - das sind bereits $50000$ Datenpaare $(t, \Delta P)$ für eine Aufnahmelänge von nur einer Sekunde! Mithilfe von Python lässt sich eine solche Datenmenge jedoch ohne Probleme verarbeiten und darstellen (siehe Grafik unten links).

Man sieht, dass es sich bei der aufgenommenen Schallwelle um eine Überlagerung von Wellen verschiedener Frequenzen handelt. Mit einem komplexen Analyseverfahren, welches sich **Fourieranalyse** nennt, lassen sich die Frequenzen der überlagerten Wellen rekonstruieren, siehe Grafik unten rechts. Ein solcher Algorithmus ist selbst für eine kleine Datenmenge von Hand sehr aufwendig durchzuführen. Somit sind Computer auch hier nicht wegzudenken.

<img align="center" src="IMG_Experiment.png" width="800" height="600">

> **Am Rande**: Eine schwingende Gitarrensaite lässt sich durch eine stehende Welle beschreiben, welche an beiden Enden fixiert ist. Solche Wellen haben ein diskretes Frequenzspektrum welches aus der Grundfrequenz $f_0$ und ihren ganzzahligen Vielfachen besteht. <font color='red'>Zusätzlich lässt sich auch zeigen, dass die Amplituden der Teilschwingungen bei steigender Frequenz mit $1/n$ abfallen.</font> Wird eine Gitarrensaite jedoch mittig (oder fast mittig) angeschlagen so fallen alle möglichen Frequenzen weg deren stehende Wellen einen Knoten in der Mitte haben. Konkret ist in unserem Beispiel die Frequenz $f_1$ dadurch stark gedämpft.

#### I.3. Beispiel: Neuronale Netzwerke

In der heutigen Forschung ist der Einsatz von neuronalen Netzwerken (bzw. Deep Learning) nicht mehr wegzudenken. Solche Netzwerke sind in der Lage komplexe Muster und Zusammenhänge zu erkennen, die daraufhin verwendet werden können um bestimmte Vorhersagen zu treffen. Ein typisches Beispiel ist die Klassifizierung von Bildern. So kann man bspw. ein neuronales Netzwerk dazu trainieren mit hoher Präzision Bilder von Hunden und Katzen zu trennen. Und das, ohne dem Netzwerk explizit mitzuteilen auf welche Merkmale hierbei zu achten ist! Ein typisches Lernschema eines neuronalen Netzwerks (NN) ist unten dargestellt. Als Training erhält das NN bereits klassifizierte Daten (Tierbilder und die zugehörige Klasse: Hund oder Katze) und nach einem ausreichend langem Trainingsprozess kann man dann Daten ohne Klassenlabel vom NN klassifizieren lassen. An dieser Stelle kann man dem Netzwerk auch Bilder übergeben die streng genommen nicht in das Schema Hund - Katze fallen, wie z.B. das Bild eines Tigers. Ein gut trainiertes Netzwerk wird den Tiger als Katze klassfizieren, da die äußerlichen Merkmale eines Tigers einer Hauskatze deutlich ähnlicher sind.

Ein konkretes Beispiel aus der Forschung wäre die Rekonstruktion von Energien in einem Teilchendetektor. Die Trainingsdaten entstehen hierbei oft durch komplexe Simulationen der Interaktionen zwischen Teilchen und Materie. 

<img align="center" src="IMG_NeuronalesNetzwerk.PNG" width="700" height="500">

### II. Python Tutorial (Teil 1/3)

#### 1. Grundlegende Rechenoperationen und Variablen
<a id="C1"></a>

In seiner simpelsten Form kann man Python als einen gewöhnlichen Taschenrechner verwenden. Der folgende Codeblock enthält einige Beispiele für einfache Rechenoperationen die man mit Python ausführen kann. Mit dem Befehl ```print()``` weist man das Program an, das Ergebnis nach der Rechnung auszugeben. Klicken Sie auf den Codeblock und dann anschließend auf *Start* (▶) in der oberen Leiste um ihn auszuführen. Unterhalb des Codeblocks werden die Ergebnisse der Rechungen angezeigt.

In [21]:
print(2 + 3)
print(2 - 3)
print(2 / 3)
print(2 * 3)

5
-1
0.6666666666666666
6


**Learning by Doing!** Wenn es um das Erlernen von Programmiersprachen geht ist es essentiell alles was man in der Theorie lernt auch direkt in die Tat umzusetzen und selbst auszuprobieren. Wir empfehlen es Ihnen alle Codebeispiele (so einfach sie auch sein mögen) die sie hier sehen nochmals in einer abgewandelten Form selbst zu schreiben. Dies kann direkt im selben Codeblock geschehen. Ergänzen Sie den oberen Codeblock mit Ihren eigenen Beispielen und probieren Sie auch Zeilen zwischen den Befehlen freizulassen, sowie die Leerzeichen in den Klammern wegzulassen. Was fällt auf?

> **Tipp:** Jeder ```print()```-Befehl in Python schreibt seinen Inhalt in eine neue Zeile. Möchte man Zeilen freilassen, so kann man einfach einen leeren ```print()```-Befehl verwenden ohne ein Argument zu übergeben.

Wir können Rechnungen in Python selbstverständlich auch auf alle reellen Zahlen erweitern. Möchte man Potenzen berechnen so schreibt man ```x**n``` für $x^n$. Führen Sie den untenstehenden Codeblock aus und achten Sie auf das Ergebnis.

In [23]:
# Dies ist ein Kommentar.
# Texte die mit einem Hashtag (#) beginnen werden von Python ignoriert.
"""
Möchte man Kommentare über mehrere Zeilen verfassen kann man den Text in drei Anführungsstriche setzen.
Kommentare eignen sich hervorragend um den eigenen Code zu erklären (für sich selbst und andere)!
Ergänzen Sie diesen Codeblock wieder mit Ihren eigenen Beispielen!
"""

print(-1.27 + 0.48)
print(2**3)
print(9**1/2)
print(9**(1/2))
print(2*10**5)
print(2e+5) 

-0.79
8
4.5
3.0
200000
200000.0


Wie man sehen kann ist auch im Programmieren Klammersetzung äußerst wichtig! ```9**1/2``` wird nämlich nicht als $9^{1/2} = \sqrt{9} = 3$ interpretiert, sondern als $9^1/2 = 4.5$. Für die wissenschaftliche Schreibweise von sehr großen, bzw. sehr kleinen Zahlen kann man abkürzend ```ae+n``` bzw. ```ae-n``` für $a \cdot 10^{\pm n}$ schreiben. Doch weshalb werden ganzen Zahlen manchmal mit einem "komma null" ausgegeben und manchmal nicht? In Python werden Zahlen in der Regel in zwei Typen unterteilt: Ganze Zahlen (*integer*) und Kommazahlen (*float*). Ist eine ganze Zahl, wie z.B. die $3$, als *float* gespeichert so wird sie immer als Kommazahl ausgegeben: ```3.0```. Eine automatische Umwandlung in einen *integer* findet nicht statt (und ist in den meisten Fällen auch nicht notwendig). Im Laufe dieser Einführung werden wir noch weitere Datentypen kennenlernen.

> **Am Rande:** Der Grund für die Trennung von Zahlen in *integer* und *float* ist von technischer Natur. Speichert man eine Zahl als *integer* ab verbraucht sie weniger Speicherplatz als wenn man sie als *float* abspeichern würde. In diesem Kurs werden Sie sich über Speicherkapazitäten allerdings keine Gedanken machen müssen. Dies wird erst dann wichtig, wenn riesige Datenmengen verarbeitet werden müssen oder Programme für Geräte entwickelt werden sollen die über sehr begrenzte Speicherkapazitäten verfügen.

Im ersten Codeblock wurde der Bruch $2/3$ berechnet und als Kommazahl mit 16 Nachkommastellen ausgegeben. Eine solche Genauigkeit ist häufig nicht nötig oder sogar unerwünscht! Mithilfe der ```round()```-Funktion können Zahlen in Python auf beliebig viele Stellen gerundet werden. Hierfür kann man der Funktion zwei Argumente liefern: Zum einen die Zahl die gerundet werden soll und zum anderen die Anzahl der Stellen auf die man runden möchte. Gibt man die Anzahl der Stellen nicht an wird immer auf eine ganze Zahl gerundet.

In [4]:
print(2 / 3)
print(round(2 / 3, 3))
print(round(2 / 3, 2))
print(round(2 / 3, 1))
print(round(2 / 3))

#Nicht vergessen: Rechnen Sie weitere Beispiele aus!

0.6666666666666666
0.667
0.67
0.7
1


Wir haben nun zwei Funktionen ineinander verschachtelt: ```print()``` und ```round()```. Wenn allerdings noch weitere Funktionen hinzukommen wird die Codezeile schnell unübersichtlich. Häufig möchte man auch einige Zahlen oder Ergebnisse an anderer Stelle wiederverwenden ohne den zugehörigen Code der das Ergebnis berechnet erneut abzutippen. Beide Probleme lassen sich mit der Benutzung von **Variablen** beheben. Betrachten Sie dazu den unten stehenden Codeblock.

In [5]:
x = 1.5
y = 2.25
z = x + y
print(z)
z = round(z)
print(z) # anstatt von print(round(z))

# Hier könnte Ihr Code stehen!

3.75
4


An diesem Beispiel werden direkt einige wichtige Syntax-Prinzipien von Python klar. Erstens ließt Python strikt **von oben nach unten**. Erst wird eine Variable ```x``` definiert und ihr der Wert $1.5$ zugewiesen. Danach wird ```y``` definiert und ihr der Wert $2.25$ zugewiesen. Nun wird ```z``` definiert und ihr der Wert $3.75 \ (= x + y)$ zugewiesen. Eine Zeile später wird der Wert von ```z``` ausgegeben $(3.75)$. Man sieht auch, dass Python Variablendefinitionen **von rechts nach links** ließt. In der vorletzten Zeile wird der Wert der Variable ```z``` $(3.75)$ gerundet $(\rightarrow 4)$ und wieder als ```z``` abgespeichert. Der Speicherplatz wurde dadurch überschrieben. Darum wird in der letzten Zeile auch nicht $3.75$, sondern $4$ ausgegeben.

> **Tipp:** Es ist häufig sehr hilfreich einen Code Zeile für Zeile durchzugehen um genau zu verstehen was für einen Zweck der Code bzw. ein Teil des Codes hat. Das ist auch besonders wichtig um nachzuvollziehen wo beim Programmieren möglicherweise ein Fehler unterlaufen ist!

> **Am Rande:** Als **Syntax** bezeichnet man die "Grammatik" einer Programmiersprache. Sie unterscheidet sich von Sprache zu Sprache. Python besitzt eine vergleichsweise sehr einfache und intuitive Syntax.

Möchte man mehrere Variablen definieren, so bietet es sich an, diese in einer Zeile zu definieren um Platz zu sparen:

In [6]:
v_x, v_y, v_z = 1.5, 2.3, 0.4 # Variablen können (fast) beliebig benannt werden.
abs_v = (v_x**2 + v_y**2 + v_z**2)**(1/2)
print(round(abs_v, 2))

2.77


Beim Benennen von Variablen sind einem nur wenig Grenzen gesetzt. Allerdings dürfen Variablennamen **nicht** mit einer Zahl beginnen und auch keine Leerzeichen enthalten:

In [27]:
1st variable = 1

SyntaxError: invalid syntax (914961040.py, line 1)

Der Compiler zeigt hier eine Fehlermeldung an die uns darauf hinweist, dass unsere Eingabe nicht den Syntax-Regeln entspricht.

> **Am Rande:** Der **Compiler** ist das Programm, das den geschriebenen Code in Maschinensprache übersetzt damit der Prozessor ihn ausführen kann.

#### 2. Listen
<a id="C2"></a>

Die Variablen ```v_x, v_y, v_z``` entsprechen in unserem Fall den Komponenten eines dreidimensionalen Vektors $\vec{v}$. In Python lassen sich zusammengehörige Variablen in **Listen** gruppieren um einfacher mit ihnen zu arbeiten. Um eine Liste zu erstellen schreibt man alle Variablen die man in die Liste einfügen möchte in eckige Klammern. Eine Liste muss auch, wie eine Variable, benannt werden:

In [7]:
v = [v_x, v_y, v_z]
w = [2.2, -1.6, -0.5] # Eine Liste kann auch direkt mit Zahlenwerten initialisiert werden.
print(v)
print(w)

[1.5, 2.3, 0.4]
[2.2, -1.6, -0.5]


Eine Liste ist eine geordnete Datenstruktur; die Elemente einer Liste besitzen also eine festgelegte Reihenfolge! So ist das erste Element der Liste ```v``` der Zahlenwert von ```v_x``` (da ```v_x``` bei der Initialisierung zuerst genannt wurde). Auf die Elemente einer Liste kann man mit dem Befehl ```v[i]``` einzeln zugreifen, wobei ```i``` der (i + 1)-te Eintrag der Liste ist. Der Zugriff auf Listenelemente kann auch rückwarts geschehen, hier wäre ```v[-1]``` der letzte Eintrag der Liste,  ```v[-2]``` der zweitletzte, usw.

In [8]:
print(w[0])
print(w[1])
print(w[2])
print(w[-1])
print(w[-2])
print(w[-3])

2.2
-1.6
-0.5
-0.5
-1.6
2.2


Eine Liste kommt mit einer ganzen Reihe von Funktionen. So kann man sich zum Beispiel die Länge einer Liste oder auch das kleinste bzw. größte Element ausgeben lassen. Elemente können nachträglich auch hinzugefügt oder gelöscht werden. Funktionen die die Liste verändern werden einfach mit ```<Liste>.<Funktion>()``` aufgerufen wie im unteren Beispiel zu sehen ist. Es entsteht somit keine veränderte Kopie der Liste welche abgespeichert werden muss; die Liste wird in ihrem Original verändert.

In [28]:
A = [0.3, 2, -0.4, 9, 3.3]

A.pop(1) # .pop(i) löscht das Element mit Index i.
print(A) # Die Variable "A" wurde damit verändert. 

A.insert(2, 0) # .insert(i, x) fügt das Element x so ein, dass es den Index i erhält.
print(A)

A.append(1.2) # .append(x) hängt das Element x am Ende der Liste an.
print(A)

print(len(A))
print(max(A))
print(min(A))

[0.3, -0.4, 9, 3.3]
[0.3, -0.4, 0, 9, 3.3]
[0.3, -0.4, 0, 9, 3.3, 1.2]
6
9
-0.4


Wir können nun die Länge des Vektors $\vec{v}$ auch wie folgt berechnen:

In [11]:
abs_v = (v[0]**2 + v[1]**2 + v[2]**2)**(1/2)
print(round(abs_v, 2))

2.77


#### 3. Schleifen
<a id="C3"></a>

In der Mathematik müssen Vektoren nicht immer dreidimensional sein. Unter der Vorraussetzung, dass die Vektorlänge ähnlich wie oben berechnet werden kann, wie könnte man die Länge eines zehndimensionalen Vektors mit minimaler Schreibarbeit berechnen? Die Antwort: **Schleifen!** Jedesmal wenn eine größere Anzahl an ähnlichen Schritten ausgeführt werden soll kann man mit Schleifen die Wiederholungen automatisieren. Eine der einfachsten Schleifen ist die *while*-Schleife. Sie wiederholt einen Satz an Anweisungen solange die **Schleifenbedingung** erfüllt ist. Hier ein einfaches Beispiel:

In [12]:
i = 0
while i < 3:
    print(i) # Schleifenanweisungen sind immer eingerückt. Dies geschieht in der Regel automatisch.
    i = i + 1

0
1
2


Wie arbeitet eine solche Schleife? Zunächst wird eine Variable ```i``` definiert und auf null gesetzt. Dann wird die Schleife initialisiert: Sie soll die Anweisungen ```print(i)``` und ```i = i + 1``` solange wiederholen, solange die Variable ```i``` kleiner als 3 ist. Im ersten Durchlauf ist ```i``` = 0; die Schleifenbedingung ist offensichtlich erfüllt und damit wird ```i``` durch den ```print()```-Befehl ausgegeben. Danach wird ```i``` um 1 erhöht. Die Schleife springt wieder an ihren Anfang und überprüft ob ```i```(= 1) kleiner 3 ist. Dies ist wieder wahr und die Anweisungen werden erneut ausgeführt. Das alles wiederholt sich solange bis ```i``` = 3 ist. Dann ist die Schleifenbedingung nicht mehr erfüllt, die Schleife wird abgebrochen und der Code ganz normal fortgesetzt.

**Achtung!** Was passiert wenn der Befehl ```i = i + 1``` nicht existieren würde? Richtig, die Schleife wird endlos weiterlaufen und Python wird immer wieder die Zahl 0 ausgeben. Sollte so etwas passieren kann das Programm über das *Stopp*-Symbol (⏹) in der oberen Leiste abgebrochen werden.

Wir haben nun ein Werkzeug gefunden um die Länge des zehndimensionalen Vektors ohne viel Schreibaufwand zu berechnen:

In [30]:
u = [2, 1.1, 0, -1.3, 0.4, -3.2, 1.8, 0, -4.3, 2]

u_sqrd, i = 0, 0
while i < len(u): # Die möglichen Indices der Liste sind 0,1,2,...,9
    u_sqrd = u_sqrd + u[i]**2
    i = i + 1

abs_u = u_sqrd**(1/2)
print(round(abs_u, 2))

6.56


Nehmen Sie sich die Zeit den obigen Code Zeile für Zeile durchzugehen und zu verstehen was in jeder Zeile passiert.

In der ersten Zeile wird der zehndimensionale Vektor initialisiert. In der zweiten Zeile wird der Index ```i``` definiert und auf Null gesetzt, allerdings auch eine Variable mit dem Namen ```u_sqrd```. Wie der Name impliziert, soll es sich hierbei um das Skalarprodukt des Vektors mit sich selbst handeln, also $\vec{u} \cdot \vec{u} = \sum_{i = 1}^{10} u_i^2$. Jeder Iterationsschritt der folgenden *while*-Schleife entspricht einem Summanden dieser Summe. Am Ende trägt die Variable ```u_sqrd``` den Wert $\vec{u} \cdot \vec{u} = |\vec{u}|^2$, entsprechend muss abschließend noch die Wurzel gezogen werden.

> **Achtung:** Es ist wichtig die Variable ```u_sqrd``` **außerhalb** der Schleife zu initialisieren. Tut man das nicht, so wird bei jedem Iterationsschritt ```u_sqrd``` wieder zurück auf Null gesetzt und man findet abschließend nur $|\vec{u}|^2 = u_{10}^2$.

Eine weitere Schleifenart ist die *for*-Schleife. Sie funktioniert sehr ähnlich wie die *while*-Schleife, jedoch mit einem Unterschied: Man muss in ihren Schleifenanweisungen nicht angeben dass der Index ```i``` sich um 1 erhöhen muss. Stattdessen gibt man bei der Initialisierung direkt an in welchem Bereich sich der Index bewegen soll und ```i``` wird automatisch erhöht:

In [14]:
u_sqrd = 0
for i in range(0, len(u)): # range(0, N) -> i soll sich sich zwischen 0 und N bewegen (exklusive N) 
    u_sqrd = u_sqrd + u[i]**2
    
abs_u = u_sqrd**(1/2)
print(round(abs_u, 2))

6.56


*for*-Schleifen eignen sich hervorragend wenn die Elemente einer Liste abgearbeitet werden sollen. So könnte man auch folgendes schreiben:

In [15]:
u_sqrd = 0
for item in u: # Iteriere über alle Elemente der Liste v, wobei "item" das aktuelle Element bezeichnet.
    u_sqrd = u_sqrd + item**2 # Der Variablenname "item" ist hier wieder beliebig gewählt worden.
    
abs_u = u_sqrd**(1/2)
print(round(abs_u, 2))

6.56


**Aufgabe 1:** Wenden Sie das bisher gelernte an und "verifizieren" Sie den Grenzwert der geometrischen Reihe

$$ \sum_{k = 0}^{\infty} \frac{1}{2^k} = 2 \ ,$$

indem Sie die Summe der ersten 1000 Summanden berechnen. Gehen Sie wie folgt vor:
1. Initialisieren Sie den anfänglichen Wert der Summe mit Null. Nennen Sie diese Variable "value".
2. Initialisieren Sie nun die *for*-Schleife mit ```range(0, N)```. Was ist $N$ in unserem Fall?
3. Schreiben Sie die Schleifenanweisung. Erhöhen Sie die Variable ```value``` in jedem Schritt um den Wert des Summanden mit Index $k$.
4. Geben Sie den Wert von ```value``` aus. Runden Sie das Ergebnis nicht!

> **Am Rande:** Selbstverständlich handelt es sich hierbei nicht um einen gültigen mathematischen Beweis.

#### 4. Funktionen
<a id="C4"></a>

Häufig möchte man nicht nur die Länge von einem Vektor, sondern gleich die Länge von einer Vielzahl an Vektoren berechnen. Um nicht jedes Mal eine neue Schleife schreiben zu müssen kann man sich **Funktionen** definieren und damit gleichzeitig für einen übersichtlicheren Code sorgen. Betrachten Sie den unten stehenden Codeblock und achten Sie auf die Syntax.

In [34]:
a = [1, 7, -2.4, 0, -1.1, 1.1, 4, 0.2, -1.8, 0.5]
b = [0, 0, 0, 1.2, -1.7, 2.9, 3.3, 2.8, 0.9, -2.4]

def abs_vector(v): # Definiere eine Funktion mit den Namen abs_vector, welche abhängig von der Variable v sein soll.
    v_sqrd = 0
    for item in v:
        v_sqrd = v_sqrd + item**2
    return v_sqrd**(1/2) # Gib die Wurzel der Variable v_sqrd zurück.

print(round(abs_vector(a), 2))
print(round(abs_vector(b), 2))

8.82
6.17


Eine Funktion wird also allgemein wie folgt definiert: ```def <Name der Funktion>(<Eingabevariablen>):```. Ähnlich wie bei Schleifen werden alle inneren Befehle eingerückt. Soll die Funktion schlussendlich eine Variable zurückgeben schreibt man: ```return <Variable die zurückgegeben werden soll>```. Allerdings ist es auch möglich Funktionen zu schreiben die gar nichts zurückgeben. Generell können alle möglichen Typen von Variablen als Eingabe- oder Rückgabe gewählt werden und es ist auch möglich Funktionen zu schreiben die mehrere Variablen entgegennehmen und/oder mehrere Variablen zurückgeben. Betrachten Sie die folgenden Beispiele:.

In [44]:
def f(x): # Diese Funktion hat keinen Rückgabewert. Stattdessen wird nur |x| über den print()-Befehl ausgegeben.
    print(abs(x))
    
f(-2)
print()

def g(x, y): # g nimmt zwei Variablen entgegen.
    alpha, beta = 2, -1
    return alpha*x**2 + beta*y**2

print(g(1, 3))
print()

def h(v): # h nimmt eine Liste entgegen und gibt einen Skalar und eine Liste innerhalb einer Liste zurück.
    e_v = []
    for item in v:
        e_v.append(item / abs_vector(v))
    return [abs_vector(v), e_v]

print(h([2, 0, 0]))

2

-7

[2.0, [1.0, 0.0, 0.0]]


Eine Variable von der eine Funktion abhängen soll (z.B. $x$) wird ausschließlich im Rahmen der Funktion definiert und verwendet. Eine solche Variable nennt man auch *lokal*. Das heißt konkret: Definieren wir die Variable $x$ auch außerhalb der Funktion (eine solche Variable heißt dann *global*), so wird die Variable durch die Funktion selbst nicht verändert. Folgendes Beispiel:

In [20]:
x = 0 # Definiere x (global) und weise ihr den Wert 0 zu.
print(x)

def g(): # Definiere g als Funktion, die allerdings von keiner Variable abhängt
    x = 1 # Setze die lokale Variable x gleich 1
    return x

print(g()) # Gib die lokale Variable aus (funktionsspezifisch)
print(x) # Gib die globale, unveränderte Variable aus

0
1
0


#### 5. Texte in Python: *strings*
<a id="C5"></a>

An dieser Stelle wollen wir einen weiteren Datentyp einwerfen: *strings*. Als einen *string* bezeichnet man alle Formen von Text, wie z.B. "Hello World!". Diese werden oft dazu verwendet die Ausgabe von Zahlenwerten übersichtlicher zu gestalten. Später werden wir Graphen von Funktionen zeichnen und *strings* verwenden um die Achsen zu beschriften. Wichtig ist, dass Text in Python in Anführungszeichen gesetzt wird um kenntlich zu machen dass es sich um einen *string* handelt!

In [45]:
print("Hello World!")

Hello World!


Oft ist es praktisch den Wert einer Variable in einen *string* einzuarbeiten. Als Platzhalter für diese Variable schreibt man ```%s``` in den *string* und gibt nach dem abschließenden Anführungszeichen an welche Variable für den Platzhalter eingesetzt werden soll. Dies geschieht mit einem ```%x```, falls ```x``` in diesem Fall die einzusetzende Variable war. Im unteren Codeblock sind zwei Beispiele gegeben.

In [46]:
x = 20
print("Der Wert der Variable x beträgt %s" %x)
y = 50
text = "x = %s, y = %s" %(x, y)
print(text)

Der Wert der Variable x beträgt 20
x = 20, y = 50


**Aufgabe 2:** Wir wollen nun ein physikalisches Beispiel betrachten indem das Gelernte wieder angewendet werden kann. Die unten angegebene Formel gibt die Gravitationsbeschleunigung auf der Oberfläche eines Planeten mit Dichte $\rho$ und Radius $R$ an.

$$ g(\rho, R) = \frac{4\pi G}{3} \cdot \rho \cdot R \ , \quad G = 6.6743 \cdot 10^{-11} \frac{\mathrm{N}\mathrm{m}^2}{\mathrm{kg}^2}$$

$G$ bezeichnet die Gravitationskonstante. Ergänzen Sie den unten stehenden Code und berechnen Sie damit die Gravitationsbeschleunigungen auf den Oberflächen der folgenden Planeten:

| Planet      | Dichte $\rho$ [g/cm^3] | Radius $R$ [km]|   
| ----------- | ---------------------- | -------------- |
| Erde        | 5,51                   | 6357           |
| Mond        | 3,34                   | 1737           |
| Mars        | 3,93                   | 3386           |
| Saturn      | 0.69                   | 58232          |
| Jupiter     | 1.33                   | 69911          |
| Kepler-1649c| 5,54                   | 6738           |

Einheiten können in Python nicht übertragen werden, sondern nur Zahlenwerte. Es ist somit immer wichtig im Hinterkopf zu behalten in welchen Einheiten man rechnet. Hier helfen Kommentare im Code!

In [ ]:
Planeten = ["Erde", ] # Liste mit Namen alle Planeten
Dichten = [5.51, ] # Liste mit allen Dichten [g/cm^3] (in der selben Reihenfolge wie die Liste mit Namen!!)
Radien = [6357, ] # Liste mit allen Radien [km]

pi = # Initialisiere Pi
G = # Initialisiere Gravitationskonstante

def g(rho, R):
    '''
    Beachten Sie, dass die gegebenen Größen nicht in SI-Einheiten gegeben sind.
    Hier wäre eine geeignete Stelle die Umrechnungen vorzunehmen.
    '''
    return 

for i in range(): # In welchem Bereich soll sich der Index i bewegen?
    g_Planet = # Berechnen Sie die Grav. Beschleunigung für Planet i.
    print("" %()) # Geben Sie die Ergebnisse in folgendem Format an: g(<Planet>) = <g> m/s^2

Die Lösung zu dieser Aufgabe findet Sie <a href="#C7">hier</a>.

#### 6. *if*-Abfragen
<a id="C6"></a>

Möchte man die Gravitationsbeschleunigung eines Planeten an einem beliebigen Punkt $P$ mit Abstand $r$ zum Planetenmittelpunkt berechnen, so muss man, je nach dem ob sich der Punkt $P$ **innerhalb** oder **außerhalb** des Planeten befindet, verschiedene Formeln verwenden:

$$g(r) = \begin{cases}
\frac{GM}{R^3}r &, \ r \leq R \\
\frac{GM}{r^2} &, \ r > R 
\end{cases}$$

Wie kann man eine solche zweigeteilte Funktion in Python implementieren? Mit sogenannten *if*-Abfragen! Diese funktionieren wie eine Schleife, jedoch führen Sie ihre Anweisungen immer nur **einmal** aus. Hier ein einfaches Beispiel:

In [47]:
x, y = -1, 1
if x < 0:
    print("x < 0")
if y < 0:
    print("y < 0")

x < 0


Will man zwei *if*-Abfragen, deren Bedingungen komplementär sind (ist die eine Bedingung erfüllt dann ist die andere automatisch nicht erfüllt und umgekehrt) hintereinander stellen kann man einfach ein ```else:``` verwenden. Dieser Befehl bezieht sich dann immer auf die vorherige *if*-Abfrage. Selbstverständlich können dann nicht beide Anweisungen durchgeführt werden.

In [50]:
if y > 0:
    print("y > 0")
else:
    print("y <= 0") # "<=" ist das Symbol für "kleiner gleich".

y > 0


Die Bedingungen die wir bei Schleifen und *if*-Abfragen verwenden nennen sich **bool(ean)s**. Diese sind logische Aussagen, die entweder **wahr** (True) oder **Falsch** (False) sein können. Sie bewerten komperative Aussagen welche sich mit "und" (&) und "oder" (|) verknüpfen lassen:

> **Am Rande:** *bools* sind ein Datentyp für sich. Sie nehmen den wenigsten Speicherplatz ein und werden immer dann verwendet wenn nur zwei Zustände gebraucht werden um ein gegebenes System zu beschreiben. Der *bool* "True" wird häufig mit einer 1 assoziiert, während "False" häufig mit der Null assoziiert wird. Diese Assoziation hat ihren Ursprung im Maschinencode, der sich nur aus Einsen und Nullen zusammensetzt.

In [53]:
print(2 > 0)
print(2 < 0)
print(1 == 4) # "==" -> Überprüfe ob zwei Zahlen gleich sind.
print(3 <= 3)

print()

print((2 > 1) & (0 < 1))
print((2 > 1) | (5 < 1))
print((0 > 1) & (1 == 1))
print((0 >= 1) | (3 < 3))

True
False
False
True

True
True
False
False


*if*-Abfragen können verwendet werden um Schleifen falls nötig früher abzubrechen. Hierzu verwendet man zusätzlich auch den ```break```-Befehl:

In [55]:
B = [0, 0, 0, -0.6, 0,  0.3]

for item in B: # Gib das erste Element der Liste B aus welches ungleich null ist.
    if item != 0: # "!=" -> ungleich.
        print(item)
        break

-0.6


Damit haben wir alle Werkzeuge zusammen um eine zweigeteilte Funktion zu definieren.

In [72]:
def g_Erde(r):
    M = 4 * pi * (Radien[0] * 1e+3)**3 * (Dichten[0] * 1e+3) / 3
    if r <= Radien[0]:
        return G * M * r * 1e+3 / (Radien[0] * 1e+3)**3
    else:
        return G * M / (r * 1e+3)**2
    
R_EM = 384400 # Abstand Erde - Mond in km
print("g_Erde(0 km) = %s m/s^2" %round(g_Erde(0), 2))
print("g_Erde(R_Erde / 2) = %s m/s^2" %round(g_Erde(Radien[0] / 2), 2))
print("g_Erde(R_Erde) = %s m/s^2" %round(g_Erde(Radien[0]), 2))
print("g_Erde(R_EM) = %s m/s^2" %round(g_Erde(R_EM), 3))

g_Erde(0 km) = 0.0 m/s^2
g_Erde(R_Erde / 2) = 4.89 m/s^2
g_Erde(R_Erde) = 9.79 m/s^2
g_Erde(R_EM) = 0.003 m/s^2


Der obige Code ist ein hervorragendes Beispiel dafür wie wichtig es ist Einheiten beim Programmieren im Auge zu behalten. Die obige Funktion ist nämlich so geschrieben, dass sie Abstände in Kilometern entgegennimmt und nicht in Metern (SI). Dies kann Verwirrung stiften wenn andere Personen den Code verstehen wollen, oder auch wenn Sie Ihren eigenen Code, den Sie vielleicht vor mehreren Monaten geschrieben haben, nochmals verwenden möchten. Kommentare können hier schnell Abhilfe schaffen!

Zusammen mit dem bereits gesammelten Wissen über Schleifen, Listen und *if*-Abfragen sind wir nun (theoretisch) auch in der Lage einen Algorithmus zu programmieren der die Einträge einer Liste sortieren kann. Der folgende Codeblock zeigt einen solchen Algorithmus; diesen komplett nachzuvollziehen ist allerdings nicht zwingend nötig, da er für Anfänger recht komplex sein kann.

In [73]:
L = [3, 9, -2, 0, -2, 6, 5, 1]

def sort_list(Unsorted):
    '''
    Eingabe: Unsortiere 1D-Liste (Zahlenwerte)
    Ausgabe: Sortierte 1D-Liste
    '''
    Sorted = [Unsorted[0]]
    for i in range(1, len(Unsorted)):
        for j in range(0, len(Sorted)):
            if Unsorted[i] <= Sorted[j]:
                Sorted.insert(j, Unsorted[i]) 
                break
            if j == len(Sorted) - 1:
                Sorted.append(Unsorted[i])
    return Sorted

print(sort_list(L))

[-2, -2, 0, 1, 3, 5, 6, 9]


Der obige Algorithmus ist eine Variante des **Insertion Sort**. Wie man sicherlich schnell einsehen kann ist es nicht praktisch einen solchen Algorithmus jedes Mal selbst zu programmieren falls der Bedarf besteht eine Liste zu sortieren. Viele wichtige Funktionen in Python existieren bereits (wie. z.B. ```round()``` oder ```len()```) und müssen nicht selbst geschrieben werden. Kein Einführungskurs der Welt kann die enorme Breite an existierenden Funktionen und sonstigen hilfreichen Inhalten abdecken, darum ist es **absolut notwendig**, dass Sie im Internet nach den Funktionen recherchieren die Sie für Ihr Programm benötigen. Das wird Ihnen eine Menge Zeit und Arbeit sparen. Wie Sie sich denken können existiert bereits auch ein Sortieralgorithmus für Listen:

In [74]:
L.sort()
print(L)

[-2, -2, 0, 1, 3, 5, 6, 9]


Viele weitere wichtige Funktionen, wie z.B. Sinus und Cosinus, aber auch Funktionen die die Möglichkeit bieten Graphen zu zeichnen befinden sich in sogenannten **Bibliotheken**. Zwei der wichtigsten Bibliotheken, *numpy* und *matplotlib*, werden wir im nächsten Notebook kennenlernen.